docs/go-openssl-compat.md

```diff
@@ -0,0 +1,87 @@
+# Compatibility of Go BoringCrypto and OpenSSL
+
+## Goal
+
+List implementation decisions taken to smooth out differences between Go BoringCrypto expectations and OpenSSL behavior.
+
+## Background
+
+Go BoringCrypto branch delegates core cryptographic algorithms to BoringSSL. Go does not support pluggable crypto backends, so in order to support BoringSSL, the BoringCrypto branch contains non-trivial modifications to several `crypto` packages. These modifications are not generic abstractions over crypto algorithms as they specifically target BoringSSL compatibility.
+
+The BoringSSL library was forked from OpenSSL 1.0.2 beta, a version that `go-crypto-openssl`supports and that already makes heavy use of the EVP interface. This fact facilitates the translation from BoringSSL to OpenSSL functions. Yet, Go BoringCrypto implements some algorithms using functions that are either deprecated in newer -but supported by us- OpenSSL versions or directly non-existing.
+
+In this situations were there is no direct mapping, we try to combine several OpenSSL functions so the security is not compromised, being the trade-off between speed and maintainability.
+
+## Implementation decisions
+
+### AES-GCM encryption in TLS mode
+
+#### Background
+
+AES-GCM with Additional Data (aka AEAD) is a symmetric block cipher whose text-book encryption inputs are:
+- initialization vector (aka nonce or iv)
+- plaintext
+- additional authenticated data
+
+And the outputs are:
+- ciphertext
+- authentication tag
+
+Go abstracts this algorithm using the `Seal` method in the [cipher.AEAD](https://pkg.go.dev/crypto/cipher#AEAD) interface:
+
+`Seal(dst, nonce, plaintext, additionalData []byte) []byte`
+
+Where dst is just a backing memory buffer to reduce allocations and the returned byte slice is the ciphertext with the authentication tag appended.
+
+When AES-GCM is used to encrypt a TLS payload (aka TLS mode), FIPS 140-2 IG A.5 ["Key/IV Pair Uniqueness * Requirements from SP 800-38D"](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf), and the general security consensus, requires constructing the IV parameter deterministically by concatenating two fields:
```

Comment:
This paragraph is describing how to construct an IV concatenating the fixed and the explicit field as per 800-38D 8.3, I think having a link to that spec here is helpful.

I guess you are confused because I do not mention the 2^64-1 limit imposed by FIPS 140-2 IG in the background section, which I should.

I will add a new paragraph mentioning it and referencing the IG 😸

Reply:
Also, notice that 800-38D 8.3 only imposes the 2^32 counter limit if the IV length is not 92bits (12 bytes), or when using RBG construction:

> The following requirement applies to all implementations that use either 1) the deterministic 
construction with IVs whose length is not 96, or 2) the RBG-based construction, for IVs of any 
length. In other words, unless an implementation only uses 96-bit IVs that are generated by the 
deterministic construction:

We only accept 12 bytes IVs, so this stricter limit does not apply to us.

Reply:
Perfect, that clears it all up, thanks!